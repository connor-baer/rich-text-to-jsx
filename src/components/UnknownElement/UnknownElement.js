import React, { Fragment } from 'react';

/**
 * Default element when the node type is unknown.
 * Renders a prominent warning in development.
 */
function UnknownElement(props) {
  if (process.env.NODE_ENV !== 'production') {
    // eslint-disable-next-line no-console
    console.debug('Node type is undefined', props);

    return (
      <Fragment>
        <span style={{ color: 'white', backgroundColor: 'red' }}>
          Node type is undefined
        </span>
        {props.children}
      </Fragment>
    );
  }

  return props.children || null;
}

if (process.env.NODE_ENV !== 'production') {
  // eslint-disable-next-line global-require
  const PropTypes = require('prop-types');

  UnknownElement.propTypes = {
    /**
     * The element children.
     */
    children: PropTypes.oneOfType([PropTypes.element, PropTypes.array]),
  };
}

/**
 * @component
 */
export default UnknownElement;
