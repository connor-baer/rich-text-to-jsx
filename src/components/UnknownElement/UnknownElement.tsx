import React, { Fragment } from 'react';

export interface UnknownElementProps {
  /**
   * The element children.
   */
  children?: JSX.Element;
}

/**
 * Default element when the node type is unknown.
 * Renders a prominent warning in development.
 */
export default function UnknownElement(
  props: UnknownElementProps,
): JSX.Element | null {
  if (process.env.NODE_ENV !== 'production') {
    // eslint-disable-next-line no-console
    console.debug('Node type is undefined', props);

    return (
      <Fragment>
        <span style={{ color: 'white', backgroundColor: 'red' }}>
          Node type is undefined
        </span>
        {props.children}
      </Fragment>
    );
  }

  return props.children || null;
}
