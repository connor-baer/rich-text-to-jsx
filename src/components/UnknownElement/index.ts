import UnknownElement from './UnknownElement';

export default UnknownElement;
