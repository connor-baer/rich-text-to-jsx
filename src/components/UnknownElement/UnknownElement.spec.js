import React from 'react';
import { create } from 'react-test-renderer';

import UnknownElement from './UnknownElement';

/* eslint-disable no-console */

describe('UnknownElement', () => {
  describe('in development', () => {
    const nodeEnv = process.env.NODE_ENV;
    const { debug } = console;

    beforeAll(() => {
      console.debug = jest.fn();
      process.env.NODE_ENV = 'development';
    });

    afterAll(() => {
      console.debug = debug;
      process.env.NODE_ENV = nodeEnv;
    });

    it('should log a debug message to the console', () => {
      create(<UnknownElement />);
      expect(console.debug).toHaveBeenCalledTimes(1);
    });

    it('should render a warning', () => {
      const actual = create(<UnknownElement />);
      expect(actual).toMatchSnapshot();
    });

    it('should render a warning and the children', () => {
      const actual = create(
        <UnknownElement>
          <p>Foo</p>
        </UnknownElement>,
      );
      expect(actual).toMatchSnapshot();
    });
  });

  describe('in production', () => {
    const nodeEnv = process.env.NODE_ENV;

    beforeAll(() => {
      process.env.NODE_ENV = 'production';
    });

    afterAll(() => {
      process.env.NODE_ENV = nodeEnv;
    });

    it('should render null', () => {
      const actual = create(<UnknownElement />);
      expect(actual).toMatchSnapshot();
    });

    it('should render the children', () => {
      const actual = create(
        <UnknownElement>
          <p>Bar</p>
        </UnknownElement>,
      );
      expect(actual).toMatchSnapshot();
    });
  });
});
