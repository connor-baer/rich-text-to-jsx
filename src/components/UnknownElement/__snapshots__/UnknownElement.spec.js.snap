// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`UnknownElement in development should render a warning 1`] = `
<span
  style="color: white; background-color: red;"
>
  Node type is undefined
</span>
`;

exports[`UnknownElement in development should render a warning and the children 1`] = `
<div>
  <span
    style="color: white; background-color: red;"
  >
    Node type is undefined
  </span>
  <p>
    Foo
  </p>
</div>
`;

exports[`UnknownElement in production should render null 1`] = `null`;

exports[`UnknownElement in production should render the children 1`] = `
<p>
  Bar
</p>
`;
